File: dart/06-io-operations/io_errors.dart
import 'dart:io';

void main() {
  final file = File('does_not_exist.txt');

  try {
    String contents = file.readAsStringSync();
    print(contents);
  } on PathNotFoundException catch (e) {
    // Catch the specific failure we expect
    print('Could not read file: ${e.path}');
  } catch (e) {
    // Catch anything else
    print('Unexpected error: $e');
  }

  // Alternatively, check before acting
  if (file.existsSync()) {
    print('File exists.');
  } else {
    print('File does not exist, skipping read.');
  }
}

File: nim/06-io-operations/io_errors.nim
import std/strutils

proc readConfig(path: string): string =
  try:
    result = readFile(path)
  except IOError as e:
    echo "Could not read ", path, ": ", e.msg
    result = ""

# This file does not exist, so readFile raises IOError
let missing = readConfig("does_not_exist.txt")
echo "Length of missing file contents: ", missing.len

# Create a file, then read it back successfully
writeFile("config.txt", "setting=on\n")
let config = readConfig("config.txt")
echo "Config: ", config.strip()
